File: 2_examples/java/src/main/java/no/systek/graphqlworkshop/storage/DataSource.java
package no.systek.graphqlworkshop.storage;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.stream.Collectors;
import org.springframework.stereotype.Component;
import static java.util.Arrays.asList;

@Component
public class DataSource {

  private final Collection<Ingredient> ingredients = new ArrayList<>();
  private final Collection<Dish> dishes = new ArrayList<>();
  private final Collection<DishIngredient> dishIngredients = new ArrayList<>();
  private final Collection<Order> orders = new ArrayList<>();

  {
    ingredients.addAll(asList(
        new Ingredient(1, "Good Fish"),
        new Ingredient(2, "Bad Fish"),
        new Ingredient(3, "Wasabi"),
        new Ingredient(4, "Soy Sauce"),
        new Ingredient(5, "Wicked Fish")
    ));

    dishes.addAll(asList(
        new Dish(100, "Sushi 8 pieces", 89.0),
        new Dish(101, "Sushi 16 pieces", 89.0)
    ));

    dishIngredients.addAll(Collections.singletonList(
        new DishIngredient(100, asList(1L, 3L))
    ));
  }

  public Collection<Dish> getDishes() {
    return dishes;
  }

  public Dish getDish(Long id) {
    return dishes.stream()
        .filter(dish -> dish.getId() == id)
        .findFirst()
        .orElseThrow(IllegalArgumentException::new);
  }

  public void addOrder(Order order) {
    orders.add(order);
  }

  public Collection<Ingredient> getIngredients() {
    return ingredients;
  }

  private Ingredient getIngredient(long id) {
    return ingredients.stream().filter(ingredient -> ingredient.getId() == id).findFirst().orElseThrow(IllegalAccessError::new);
  }

  public Collection<Ingredient> findIngredientsInDish(long dishId) {
    Collection<Long> ingredientIds = dishIngredients.stream()
        .filter(dishIngredient -> dishIngredient.getDishId() == dishId)
        .findFirst()
        .orElseThrow(IllegalArgumentException::new).getIngredientIds();

    return ingredientIds.stream().map(this::getIngredient).collect(Collectors.toList());
  }

  public Collection<Order> getOrders() {
    return orders;
  }
}

File: 3_examples/kotlin/src/main/kotlin/no/systek/graphqlworkshop/storage/DataSource.kt
package no.systek.graphqlworkshop.storage

import org.springframework.stereotype.Component

@Component
class DataSource {
    final val ingredients: MutableCollection<Ingredient>
    final val dishes: MutableCollection<Dish>
    final val dishIngredients: MutableCollection<DishIngredient>
    final val orders: MutableCollection<Order> = mutableListOf()

    init {
        ingredients = mutableListOf(
                Ingredient(1, "Good Fish"),
                Ingredient(2, "Bad Fish"),
                Ingredient(3, "Wasabi"),
                Ingredient(4, "Soy Sauce"),
                Ingredient(5, "Wicked Fish"))

        dishes = mutableListOf(
                Dish(100, "Sushi 8 pieces", 89.0),
                Dish(101, "Sushi 16 pieces", 89.0)
        )
        dishIngredients = mutableListOf(
                DishIngredient(100, listOf(1, 3))
        )
    }

    fun getDish(id: Long) = dishes.first { it.id == id }

    fun getIngredient(id: Long) = ingredients.first { it.id == id }

    fun addOrder(order: Order) = orders.add(order)

    fun findIngredientsInDish(dishId: Long): Collection<Ingredient> {
        val ingredientIds = dishIngredients.firstOrNull { it.dishId == dishId }?.ingredientIds ?: return emptyList()

        return ingredients.filter { it.id in ingredientIds }
    }
}
